builtins/web/file.cpp

```diff
@@ -122,45 +107,47 @@ bool File::lastModified_get(JSContext *cx, unsigned argc, JS::Value *vp) {
     return api::throw_error(cx, api::Errors::WrongReceiver, "lastModified get", "File");
   }
 
-  auto lastModified = JS::GetReservedSlot(self, static_cast<size_t>(File::Slots::LastModified)).toInt32();
+  auto lastModified =
+      JS::GetReservedSlot(self, static_cast<size_t>(Slots::LastModified)).toNumber();
   args.rval().setNumber(lastModified);
   return true;
 }
 
-bool File::init_last_modified(JSContext *cx, HandleObject self, HandleValue initv) {
-  bool has_last_modified = false;
-
-  JS::RootedValue init_val(cx, initv);
-  JS::RootedObject opts(cx, init_val.toObjectOrNull());
-
-  if (!opts) {
-    return true;
+bool File::init(JSContext *cx, HandleObject self, HandleValue fileBits, HandleValue fileName, HandleValue opts) {
+  // 1. Let bytes be the result of processing blob parts given fileBits and options.
+  if (!Blob::init(cx, self, fileBits, opts)) {
+    return false;
   }
 
-  if (!JS_HasProperty(cx, opts, "lastModified", &has_last_modified)) {
+  // 2. Let n be the fileName argument to the constructor.
+  RootedString name(cx, JS::ToString(cx, fileName));
+  if (!name) {
     return false;
   }
 
-  if (has_last_modified) {
-    JS::RootedValue ts(cx);
-    if (!JS_GetProperty(cx, opts, "lastModified", &ts)) {
-      return false;
-    }
-
-    if (ts.isNumber()) {
-      SetReservedSlot(self, static_cast<uint32_t>(Slots::LastModified), ts);
-    }
+  // 3. Process `FilePropertyBag` dictionary argument by running the following substeps:
+  //  1. and 2 - the steps for processing a `type` member are ensured by Blob implementation.
+  //  3. If the `lastModified` member is provided, let d be set to the lastModified dictionary
+  //  member. If it is not provided, set d to the current date and time represented as the number of
+  //  milliseconds since the Unix Epoch.
+  RootedValue lastModified(cx);
+  if (!init_last_modified(cx, opts, &lastModified)) {
+    return false;
   }
 
-  return true;
-}
-
-JSObject *File::blob(JSObject *self) {
-  MOZ_ASSERT(is_instance(self));
-  auto blob = &JS::GetReservedSlot(self, static_cast<size_t>(Blob::Slots::Data)).toObject();
+  // Return a new File object F such that:
+  //  2. F refers to the bytes byte sequence.
+  //  3. F.size is set to the number of total bytes in bytes.
+  //  4. F.name is set to n.
+  //  5. F.type is set to t.
+  //  6. F.lastModified is set to d.
+  //
+  //  Steps 2, 3 and 5 are handled by Blob. We extend the Blob by adding a `name`
+  //  and the `lastModified` properties.
+  SetReservedSlot(self, static_cast<uint32_t>(Slots::Name), JS::StringValue(name));
+  SetReservedSlot(self, static_cast<uint32_t>(Slots::LastModified), lastModified);
```

Comment:
This part is a bit unfortunate: we can't store `lastModified` as a `uint32_t`, because that'd overflow very quickly. (Way way before 2038, because it includes milliseconds.)

I think the best thing to do so we don't have to step away from using a `Value` encoding is to use a `Number`: 53 bits of precision should still be plenty for our purposes here.